File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.py

def solution(n, m, x, y, r, c, k):
    dis = abs(x-r)+abs(y-c)
    if dis%2 != k%2 or dis>k:
        return 'impossible'
    
    answer = []
    d,l,r,u,o = max(0,r-x), max(0,y-c),max(0,c-y),max(0,x-r), (k-dis)//2

    for i in range(k):
        if x<n and d:
            answer.append('d')
            x+=1
            d-=1
            continue
        if x<n and o:
            answer.append('d')
            x+=1
            u+=1
            o-=1
            continue
        
        if y>1 and o:
            answer.append('l')
            y-=1
            r+=1
            o-=1
            continue
        if y>1 and l:
            answer.append('l')
            y-=1
            l-=1
            continue
            
        if y<m and r:
            answer.append('r')
            y+=1
            r-=1
            continue  
        if y<m and o:
            answer.append('r')
            y+=1
            l+=1
            o-=1
            continue
        if x>1 and u:
            answer.append('u')
            x-=1
            u-=1
            continue
        if x>1 and o:
            answer.append('u')
            x-=1
            d+=1
            o-=1
            continue

        
    
                
                
            

    
        
    return ''.join(answer)

File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.java
public class Solution {
    public static String solution(int n, int m, int x, int y, int r, int c, int k) {
        int dis = Math.abs(x - r) + Math.abs(y - c);
        if (dis % 2 != k % 2 || dis > k) {
            return "impossible";
        }

        StringBuilder answer = new StringBuilder();
        int d = Math.max(0, r - x);
        int l = Math.max(0, y - c);
        int rr = Math.max(0, c - y); // r과 변수명이 겹쳐서 rr로 변경
        int u = Math.max(0, x - r);
        int o = (k - dis) / 2;

        for (int i = 0; i < k; i++) {
            if (x < n && d > 0) {
                answer.append('d');
                x++;
                d--;
                continue;
            }
            if (x < n && o > 0) {
                answer.append('d');
                x++;
                u++;
                o--;
                continue;
            }

            if (y > 1 && o > 0) {
                answer.append('l');
                y--;
                rr++;
                o--;
                continue;
            }
            if (y > 1 && l > 0) {
                answer.append('l');
                y--;
                l--;
                continue;
            }

            if (y < m && rr > 0) {
                answer.append('r');
                y++;
                rr--;
                continue;
            }
            if (y < m && o > 0) {
                answer.append('r');
                y++;
                l++;
                o--;
                continue;
            }

            if (x > 1 && u > 0) {
                answer.append('u');
                x--;
                u--;
                continue;
            }
            if (x > 1 && o > 0) {
                answer.append('u');
                x--;
                d++;
                o--;
                continue;
            }
        }

        return answer.toString();
    }

    public static void main(String[] args) {
        // 예시 실행
        System.out.println(solution(3, 4, 2, 3, 3, 1, 5));
    }
}
